File: module-10-fine-tuning/solutions/dpo-dataset.ts
/**
 * Module 10: DPO dataset builder
 * Build (prompt, preferred, dispreferred) pairs from eval results.
 */
import * as fs from "fs/promises";
import type { TraceRecord } from "./data-pipeline.js";

export interface DPOExample {
  prompt: string;
  chosen: string; // preferred (successful) completion
  rejected: string; // dispreferred (failed) completion
}

/** Build DPO pairs by matching successful vs failed trials for same task type */
export function buildDPODataset(traces: TraceRecord[]): DPOExample[] {
  // Group by task
  const byTask = new Map<string, TraceRecord[]>();
  for (const trace of traces) {
    const key = trace.task;
    if (!byTask.has(key)) byTask.set(key, []);
    byTask.get(key)!.push(trace);
  }

  const pairs: DPOExample[] = [];

  for (const [task, taskTraces] of byTask) {
    const successes = taskTraces.filter((t) => t.success);
    const failures = taskTraces.filter((t) => !t.success);

    // Pair each success with each failure
    for (const success of successes) {
      for (const failure of failures) {
        pairs.push({
          prompt: task,
          chosen: formatTrace(success),
          rejected: formatTrace(failure),
        });
      }
    }
  }

  console.log(`Built ${pairs.length} DPO pairs from ${traces.length} traces`);
  return pairs;
}

function formatTrace(trace: TraceRecord): string {
  return trace.transcript.turns
    .filter((t) => t.role === "assistant")
    .map((t) => {
      let content = t.content;
      if (t.toolCalls?.length) {
        content += "\n" + t.toolCalls.map((tc) => `[${tc.name}] ${tc.result.slice(0, 100)}`).join("\n");
      }
      return content;
    })
    .join("\n---\n");
}

export async function saveDPODataset(pairs: DPOExample[], outputPath: string): Promise<void> {
  const lines = pairs.map((p) => JSON.stringify(p)).join("\n");
  await fs.writeFile(outputPath, lines, "utf-8");
  console.log(`Saved ${pairs.length} DPO pairs to ${outputPath}`);
}

File: module-10-fine-tuning/solutions/python/dpo_dataset.py
"""Module 10: DPO dataset builder."""
import json


def _format_trace(trace: dict) -> str:
    parts = []
    for turn in trace["transcript"].turns:
        if turn.role == "assistant":
            content = turn.content
            if turn.tool_calls:
                content += "\n" + "\n".join(
                    f"[{tc['name']}] {tc['result'][:100]}" for tc in turn.tool_calls
                )
            parts.append(content)
    return "\n---\n".join(parts)


def build_dpo_dataset(traces: list[dict]) -> list[dict]:
    by_task: dict[str, list[dict]] = {}
    for trace in traces:
        by_task.setdefault(trace["task"], []).append(trace)

    pairs = []
    for task, task_traces in by_task.items():
        successes = [t for t in task_traces if t["success"]]
        failures = [t for t in task_traces if not t["success"]]
        for s in successes:
            for f in failures:
                pairs.append({
                    "prompt": task,
                    "chosen": _format_trace(s),
                    "rejected": _format_trace(f),
                })

    print(f"Built {len(pairs)} DPO pairs from {len(traces)} traces")
    return pairs


def save_dpo_dataset(pairs: list[dict], output_path: str) -> None:
    with open(output_path, "w") as f:
        for p in pairs:
            f.write(json.dumps(p) + "\n")
    print(f"Saved {len(pairs)} DPO pairs to {output_path}")
